### PR TITLE
Reduction of RMG-generated models

### PR DESCRIPTION
**Pull request by nickvandewiele**
Do not merge yet. PR meant for code review primarily.

Git history of development versions can be found here: 
https://github.com/nickvandewiele/reduction

Features:
- Rate-based reaction elimination of RMG-generated models
- Supports multiple reaction systems
- integrates seamlessly with RMG-Py, and rate expressions supported by RMG-Py.
- supports parallelization using SCOOP
- support for multiple constraints in the form of target observables

Todo:
- [ ] support for chemkin models without species dictionaries
- [x] support for non-parallel runs.
- [x] speed up
- [x] refactor reduction module into algorithm, I/O, and a client
- [x] writes a chemkin compatible reduced model to disk

### CODE REVIEW

**Comment by nickvandewiele** [comment]
The new bisection optimization algorithm drastically speeds up the model reduction.

File: rmgpy/reduction/main.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    atol, rtol = rmg.absoluteTolerance, rmg.relativeTolerance
+    index = 0
+    reactionSystem = rmg.reactionSystems[index]    
+    
+    #compute original target observables
+    observables = computeObservables(targets, reactionModel, reactionSystem, \
+     rmg.absoluteTolerance, rmg.relativeTolerance)
+
+    logger.info('Observables of original model:'.format())
```
Comment (KEHANG) [inline rmgpy/reduction/main.py]: did you miss putting something inside `format()`?

**Comment by KEHANG** [comment]
probably need to upload a chem_annotated.inp file in `examples/reduction/minimal/chemkin` folder to make the example work.

File: rmgpy/solver/baseTest.py [hunk 2]
```diff
@@ -101,5 +101,20 @@ def test_listen(self):
...
+        self.inputFile = working_dir + 'input.py'
+        self.chemkinFile = working_dir + 'chemkin/chem.inp'
+        self.spc_dict = working_dir + 'chemkin/species_dictionary.txt'
+        self.rmg = loadRMGPyJob(self.inputFile, self.chemkinFile, self.spc_dict)
+        self.rxnSys = self.rmg.reactionSystems[0]
+
+    def test_pickle(self):
+        import pickle
+        rxnSys = pickle.loads(pickle.dumps(self.rxnSys))
+
```
Comment (KEHANG) [inline rmgpy/solver/baseTest.py]: I think `test_pickle` should test whether the loaded rxnSys is same as original rxnSys.

**Comment by nickvandewiele** [comment]
I updated the tests in rmgpy.solver.baseTest, including the test_pickle unit test you mentioned.

**Comment by rwest** [comment]
Looks great, Nick. When can we have a seminar where you explain it to us? Also for your scoop and parallelization work. Must happen before you leave!!

**Pull request merged**